daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversionStackSafe.scala

```diff
@@ -0,0 +1,368 @@
+        val newRemapsF: Map[Int, target.SELoc] = fvs.zipWithIndex.map { case (orig, i) =>
+          (orig + arity) -> target.SELocF(i)
+        }.toMap
+        val newRemapsA = (1 to arity).map { case i =>
+          i -> target.SELocA(arity - i)
+        }
+        // The keys in newRemapsF and newRemapsA are disjoint
+        val m1 = newRemapsF ++ newRemapsA
+        Env(m1)
+      }
+    }
+
+    /** Closure-conversion, Traversal:
+      *
+      *   To ensure stack-safety, the input expression is traversed by a single tail-recursive 'loop'.
+      *   During the 'Traversal', we are either:
+      *   - going 'Down' a source expression (subtree), with an 'Env' for context, or
+      *   - coming 'Up' with a target expression (result for a subtree)
+      *
+      *   In both cases we have a continuation ('Cont') argument which says how to proceed.
+      */
+    sealed trait Traversal
+    object Traversal {
+      final case class Down(exp: source.SExpr, env: Env) extends Traversal
+      final case class Up(exp: target.SExpr) extends Traversal
+    }
+    import Traversal._
+
+    /** Closure Conversion, Cont:
+      *
+      *   The multiple forms for a continuation describe how the result of transforming a
+      *   sub-expression should be embedded in the continuing traversal. The continuation
+      *   forms correspond to the source expression forms: specifically, the location of
+      *   recursive expression instances.
+      *
+      *   For expression forms with no recursive instance (i.e. SEVar, SEVal), there are
+      *   no corresponding continuation forms.
+      *
+      *   For expression forms with a single recursive instance (i.e. SELocation), there
+      *   is a single continuation form: (Cont.Location).
+      *
+      *   For expression forms with two recursive instances (i.e. SETryCatch), there are
+      *   two corresponding continuation forms: (Cont.TryCatch1, Cont.TryCatch2).
+      *
+      *   For the more complex expression forms containinga list of recursive instances
+      *   (i.e. SEAppGeneral), the corresponding continuation forms are also more complex,
+      *   but will generally have two cases (i.e. Cont.App1, Cont.App2), corresponding to
+      *   the Nil/Cons cases of the list of recursive instances.
+      *
+      *   And so on. In effect, 'Cont' is a zipper type for expressions.
+      *
+      *   Another way to understand the continuation forms is by observing the presence of
+      *   an 'env: Env' component indicates more source-expression processing to be done
+      *   (generally with the components following the 'env'). Any components before the
+      *   'env' (or all components if there is no 'env') represent transform-(sub)-results
+      *   which need combining into the final result.
+      *
+      *   All continuation forms except 'Cont.Finish` embed a further continuation, which
+      *   describes the continuing traversal.
+      */
+    sealed trait Cont
```

Comment:
```suggestion
    private[this] abstract class Cont
```

1. If it does not need to by public, it should be private
2. abstract class are in general slightly more efficient.

Reply:
It can't be marked `private` because we inside an enclosing `def`. (And of source that is the most private!)
I changed from `trait` to `abstract class`
But it must still be a `sealed abstract class` so we get exhaustivity checking of pattern matches

Reply:
Sorry.